DemoAppPush/NotificationService.swift

```diff
@@ -0,0 +1,115 @@
+import UserNotifications
+
+class NotificationService: UNNotificationServiceExtension {
+    var contentHandler: ((UNNotificationContent) -> Void)?
+    var request: UNNotificationRequest?
+
+    func addAttachments(
+        url: URL,
+        content: UNMutableNotificationContent,
+        identifier: String = "image",
+        completion: @escaping (UNMutableNotificationContent) -> Void
+    ) {
+        let task = URLSession.shared.downloadTask(with: url) { (downloadedUrl, _, _) in
+            defer {
+                completion(content)
+            }
+
+            guard let downloadedUrl = downloadedUrl else {
+                return
+            }
+          
+            guard let path = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first else {
+                return
+            }
+          
+            let localURL = URL(fileURLWithPath: path).appendingPathComponent(url.lastPathComponent)
+          
+            do {
+                try FileManager.default.moveItem(at: downloadedUrl, to: localURL)
+            } catch {
+                return
+            }
+
+            guard let attachment = try? UNNotificationAttachment(identifier: identifier, url: localURL, options: nil) else {
+                return
+            }
+
+            content.attachments = [attachment]
+        }
+        task.resume()
+    }
+
+    func addMessageAttachments(
+        message: ChatMessage,
+        content: UNMutableNotificationContent,
+        completion: @escaping (UNMutableNotificationContent) -> Void
+    ) {
+        if let imageURL = message.author.imageURL {
+            addAttachments(url: imageURL, content: content) {
+                completion($0)
+            }
+            return
+        }
+        if let attachment = message.imageAttachments.first {
+            addAttachments(url: attachment.imageURL, content: content) {
+                completion($0)
+            }
+            return
+        }
+    }
+
+    override func didReceive(
+        _ request: UNNotificationRequest,
+        withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void
+    ) {
+        self.contentHandler = contentHandler
+        self.request = request
+
+        guard let content = request.content.mutableCopy() as? UNMutableNotificationContent else {
+            return
+        }
+
+        guard let userId = UserDefaults.standard.string(forKey: currentUserIdRegisteredForPush), let userCredentials = UserCredentials.builtInUsersByID(id: userId) else {
+            return
+        }
+
+        var config = ChatClientConfig(apiKey: .init(apiKeyString))
+        config.isLocalStorageEnabled = true
+        config.applicationGroupIdentifier = applicationGroupIdentifier
+
+        let client = ChatClient(config: config)
+        client.setToken(token: Token(stringLiteral: userCredentials.token))
+
+        let chatHandler = ChatRemoteNotificationHandler(client: client, content: content)
```

Comment:
IMHO it'd be better if this entity was able to handle any content, so it wouldn't accept `content` in `init` but it'd accept `content` in `handleNotification` as `handleNotification(content:)`. Then we'd expose the default handler from `ChatClient` as `client.remoteNotificationHandler`